src/stores/widgets/WidgetLayoutStore.ts

```diff
@@ -0,0 +1,476 @@
+    index?: number;
+
+    // Percentage (integer) for relative width of the container to consume.
+    // Clamped to 0-100 and may have minimums imposed upon it. Only applies
+    // to containers which support inner resizing (currently only the top
+    // container).
+    width?: number;
+
+    // Percentage (integer) for relative height of the container. Note that
+    // this only applies to the top container currently, and that container
+    // will take the highest value among widgets in the container. Clamped
+    // to 0-100 and may have minimums imposed on it.
+    height?: number;
+
+    // TODO: [Deferred] Maximizing (fullscreen) widgets by default.
+}
+
+interface IWidgetLayouts {
+    [widgetId: string]: IStoredLayout;
+}
+
+interface ILayoutStateEvent {
+    // TODO: [Deferred] Forced layout (fixed with no changes)
+
+    // The widget layouts.
+    widgets: IWidgetLayouts;
+}
+
+interface ILayoutSettings extends ILayoutStateEvent {
+    overrides?: string; // event ID for layout state event, if present
+}
+
+// Dev note: "Pinned" widgets are ones in the top container.
+export const MAX_PINNED = 3;
+
+// These two are whole percentages and don't really mean anything. Later values will decide
+// minimum, but these help determine proportions during our calculations here. In fact, these
+// values should be *smaller* than the actual minimums imposed by later components.
+const MIN_WIDGET_WIDTH_PCT = 10; // 10%
+const MIN_WIDGET_HEIGHT_PCT = 2; // 2%
+
+export class WidgetLayoutStore extends ReadyWatchingStore {
+    private static internalInstance: WidgetLayoutStore;
+
+    private byRoom: {
+        [roomId: string]: {
+            // @ts-ignore - TS wants a string key, but we know better
+            [container: Container]: {
+                ordered: IApp[];
+                height?: number;
+                distributions?: number[];
+            };
+        };
+    } = {};
+
+    private pinnedRef: string;
+    private layoutRef: string;
+
+    private constructor() {
+        super(defaultDispatcher);
+    }
+
+    public static get instance(): WidgetLayoutStore {
+        if (!WidgetLayoutStore.internalInstance) {
+            WidgetLayoutStore.internalInstance = new WidgetLayoutStore();
+        }
+        return WidgetLayoutStore.internalInstance;
+    }
+
+    public static emissionForRoom(room: Room): string {
+        return `update_${room.roomId}`;
+    }
+
+    private emitFor(room: Room) {
+        this.emit(WidgetLayoutStore.emissionForRoom(room));
+    }
+
+    protected async onReady(): Promise<any> {
+        this.updateAllRooms();
+
+        this.matrixClient.on("RoomState.events", this.updateRoomFromState);
+        this.pinnedRef = SettingsStore.watchSetting("Widgets.pinned", null, this.updateFromSettings);
+        this.layoutRef = SettingsStore.watchSetting("Widgets.layout", null, this.updateFromSettings);
+        WidgetStore.instance.on(UPDATE_EVENT, this.updateAllRooms);
```

Comment:
Can the `WidgetStore` emit events per room instead? I'd like to avoid calling `updateAllRooms` wherever possible. It's a lot of work to do for a large account. If we can get an event per room here, then we only need to calculate that room.

Reply:
`WidgetStore` does emit per room, but that's a lot of rooms to keep track of. Large accounts shouldn't see a large impact from this, but if it's still worrisome then I can go hack on `WidgetStore` a bit.